packages/dd-trace/src/latency_stats.js

```diff
@@ -0,0 +1,156 @@
+const os = require('os')
+const pkg = require('../../../package.json')
+const { decodePathwayContext } = require('../../datadog-plugin-kafkajs/src/hash')
+
+const { LogCollapsingLowestDenseDDSketch } = require('@datadog/sketches-js')
+
+const { LatencyStatsExporter } = require('./exporters/latency-stats')
+
+class AggStats {
+  constructor (aggKey) {
+    this.hash = aggKey.hash
+    this.parentHash = aggKey.parentHash
+    this.edgeTags = aggKey.edgeTags
+    this.edgeLatency = new LogCollapsingLowestDenseDDSketch(0.00775)
+    this.pathwayLatency = new LogCollapsingLowestDenseDDSketch(0.00775)
+  }
+
+  record (checkpoint) {
+    const edgeLatency = checkpoint.metrics.edge_latency
+    const pathwayLatency = checkpoint.metrics.pathway_latency
+    this.edgeLatency.accept(edgeLatency)
+    this.pathwayLatency.accept(pathwayLatency)
+  }
+
+  toJSON () {
+    return {
+      Hash: this.hash,
+      ParentHash: this.parentHash,
+      EdgeTags: this.edgeTags,
+      EdgeLatency: this.edgeLatency.toProto(), // TODO: custom proto encoding
+      PathwayLatency: this.pathwayLatency.toProto() // TODO: custom proto encoding
+    }
+  }
+}
+
+class AggKey {
+  constructor (checkpoint) {
+    this.hash = decodePathwayContext(checkpoint.metrics['dd-pathway-ctx'])[0]
+    this.parentHash = checkpoint.metrics.parent_hash
+    this.edgeTags = checkpoint.metrics.edge_tags
+  }
+
+  toString () {
+    return [
+      this.hash.toString(),
+      this.parentHash.toString()
+    ].join(',')
+  }
+}
+
+class SpanBuckets extends Map {
+  forCheckpoint (checkpoint) {
+    const aggKey = new AggKey(checkpoint)
+    const key = aggKey.toString()
+    // also include parentHash, edgeTags in ddsketch
+    if (!this.has(key)) {
+      this.set(key, new AggStats(aggKey)) // StatsPoint
+    }
+
+    return this.get(key)
+  }
+}
+
+class TimeBuckets extends Map {
+  forTime (time) {
+    if (!this.has(time)) {
+      this.set(time, new SpanBuckets())
+    }
+
+    return this.get(time)
+  }
+}
+
+class LatencyStatsProcessor {
+  constructor ({
+    dsmEnabled,
```

Comment:
Is this file going to be useful to features other than DSM? The location and name makes me think so. For that reason it would probably be better to have the constructor accept a property named `enabled` instead of `dsmEnabled`. I suspect the property name in this class is just due to the shape of the config object used outside of the class.

Here's one way to reconfigure the calling location without too much effort:

```js
new LatencyStatsProcessor({
  enabled: this._tracerConfig.dsmEnabled,
  ...this._tracerConfig
})
```

It does still pass in a field named `dsmEnabled` but it's then ignored by the new constructor.

Reply:
I'm not sure if it will be used by anything other than DSM though, maybe I should just move the file for now